<a href='http://moncoachdata.com'><img src= 'https://moncoachdata.com/wp-content/uploads/MonCoachData-cours/MonCoachData.jpg' width=400/></a>


---


<center><em>Copyright MonCoachData (tous droits réservés)</em></center>
<center><em>Pour plus d'informations, visitez notre site <a href='http://moncoachdata.com'>moncoachdata.com</a></em></center>



---

# Traitement des Données Catégorielles

De nombreux modèles de Machine Learning ne peuvent pas traiter les ensembles de données catégorielles comme des chaînes de caractères String. Par exemple, la régression linéaire ne peut pas appliquer un coefficient bêta à des couleurs comme "rouge" ou "bleu". Au lieu de cela, nous devons convertir ces catégories en variables "dummy" (factices), autrement connu sous le nom d'encodage "one-hot".

## Importations

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Données

Nous allons ouvrir le fichier CSV qui a été "nettoyé" pour éliminer les valeurs aberrantes (outliers) et les NaN (valeurs manquantes) à partir des vidéos précédentes.

In [ ]:
df = pd.read_csv("Ames_NO_Missing_Data.csv")

In [ ]:
df.head()

MS SubClass MS Zoning  Lot Frontage  ...  Sale Type Sale Condition SalePrice
0           20        RL         141.0  ...        WD          Normal    215000
1           20        RH          80.0  ...        WD          Normal    105000
2           20        RL          81.0  ...        WD          Normal    172000
3           20        RL          93.0  ...        WD          Normal    244000
4           60        RL          74.0  ...        WD          Normal    189900

[5 rows x 76 columns]

## Description des Données

In [ ]:
with open('Ames_Housing_Feature_Description.txt','r') as f: 
    print(f.read())

MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

MSZoning: Identifies the general zoning classification of the sale.
		
       A	Agriculture
       C	Commercial
       FV	Floating Village Residential
       I	Industrial
       RH	Residential High Density
       RL	Residential Low Density
       RP	Residential Low Density Park 
       RM

## Colonne numérique à catégorielle

Nous devons être prudents lorsqu'il s'agit d'encoder des catégories sous forme de nombres. Nous voulons nous assurer que la relation numérique a un sens pour un modèle. Par exemple, l'encodage de "MSSubClass" est essentiellement un code numérique par classe :

    MSSubClass: Identifies the type of dwelling involved in the sale.	

        20	1-STORY 1946 & NEWER ALL STYLES
        30	1-STORY 1945 & OLDER
        40	1-STORY W/FINISHED ATTIC ALL AGES
        45	1-1/2 STORY - UNFINISHED ALL AGES
        50	1-1/2 STORY FINISHED ALL AGES
        60	2-STORY 1946 & NEWER
        70	2-STORY 1945 & OLDER
        75	2-1/2 STORY ALL AGES
        80	SPLIT OR MULTI-LEVEL
        85	SPLIT FOYER
        90	DUPLEX - ALL STYLES AND AGES
       120	1-STORY PUD (Planned Unit Development) - 1946 & NEWER
       150	1-1/2 STORY PUD - ALL AGES
       160	2-STORY PUD - 1946 & NEWER
       180	PUD - MULTILEVEL - INCL SPLIT LEV/FOYER
       190	2 FAMILY CONVERSION - ALL STYLES AND AGES

Le nombre ne semble pas avoir de relation avec les autres nombres. Si 30 > 20 est vrai, il n'est pas vraiment logique que "1-STORY 1945 & OLDER" > "1-STORY 1946 & NEWER ALL STYLES". Gardez à l'esprit que ce n'est pas toujours le cas, par exemple les sièges de 1ère classe par rapport aux sièges de 2ème classe encodés avec les chiffres 1 et 2. Assurez-vous de bien comprendre votre ensemble de données pour examiner ce qui doit être converti/changé.

### MSSubClass

In [ ]:
# Convertir en String
df['MS SubClass'] = df['MS SubClass'].apply(str)

# Création de variables "Dummy"

### Éviter la multicollinéarité et le piège des variables fictives

https://stats.stackexchange.com/questions/144372/dummy-variable-trap

In [ ]:
person_state =  pd.Series(['Dead','Alive','Dead','Alive','Dead','Dead'])

In [ ]:
person_state

0     Dead
1    Alive
2     Dead
3    Alive
4     Dead
5     Dead
dtype: object

In [ ]:
pd.get_dummies(person_state)

Alive  Dead
0      0     1
1      1     0
2      0     1
3      1     0
4      0     1
5      0     1

In [ ]:
pd.get_dummies(person_state,drop_first=True)

Dead
0     1
1     0
2     1
3     0
4     1
5     1

## Création de variables Dummy à partir des objets Colonne

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html

In [ ]:
df.select_dtypes(include='object')

MS SubClass MS Zoning Street  ... Paved Drive Sale Type Sale Condition
0             20        RL   Pave  ...           P       WD          Normal
1             20        RH   Pave  ...           Y       WD          Normal
2             20        RL   Pave  ...           Y       WD          Normal
3             20        RL   Pave  ...           Y       WD          Normal
4             60        RL   Pave  ...           Y       WD          Normal
...          ...       ...    ...  ...         ...       ...            ...
2920          80        RL   Pave  ...           Y       WD          Normal
2921          20        RL   Pave  ...           Y       WD          Normal
2922          85        RL   Pave  ...           Y       WD          Normal
2923          20        RL   Pave  ...           Y       WD          Normal
2924          60        RL   Pave  ...           Y       WD          Normal

[2925 rows x 40 columns]

In [ ]:
df_nums = df.select_dtypes(exclude='object')
df_objs = df.select_dtypes(include='object')

In [ ]:
df_nums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 36 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Lot Frontage     2925 non-null   float64
 1   Lot Area         2925 non-null   int64  
 2   Overall Qual     2925 non-null   int64  
 3   Overall Cond     2925 non-null   int64  
 4   Year Built       2925 non-null   int64  
 5   Year Remod/Add   2925 non-null   int64  
 6   Mas Vnr Area     2925 non-null   float64
 7   BsmtFin SF 1     2925 non-null   float64
 8   BsmtFin SF 2     2925 non-null   float64
 9   Bsmt Unf SF      2925 non-null   float64
 10  Total Bsmt SF    2925 non-null   float64
 11  1st Flr SF       2925 non-null   int64  
 12  2nd Flr SF       2925 non-null   int64  
 13  Low Qual Fin SF  2925 non-null   int64  
 14  Gr Liv Area      2925 non-null   int64  
 15  Bsmt Full Bath   2925 non-null   float64
 16  Bsmt Half Bath   2925 non-null   float64
 17  Full Bath     

In [ ]:
df_objs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2925 entries, 0 to 2924
Data columns (total 40 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   MS SubClass     2925 non-null   object
 1   MS Zoning       2925 non-null   object
 2   Street          2925 non-null   object
 3   Lot Shape       2925 non-null   object
 4   Land Contour    2925 non-null   object
 5   Utilities       2925 non-null   object
 6   Lot Config      2925 non-null   object
 7   Land Slope      2925 non-null   object
 8   Neighborhood    2925 non-null   object
 9   Condition 1     2925 non-null   object
 10  Condition 2     2925 non-null   object
 11  Bldg Type       2925 non-null   object
 12  House Style     2925 non-null   object
 13  Roof Style      2925 non-null   object
 14  Roof Matl       2925 non-null   object
 15  Exterior 1st    2925 non-null   object
 16  Exterior 2nd    2925 non-null   object
 17  Mas Vnr Type    2925 non-null   object
 18  Exter Qu

### Conversion

In [ ]:
df_objs = pd.get_dummies(df_objs,drop_first=True)

In [ ]:
final_df = pd.concat([df_nums,df_objs],axis=1)

In [ ]:
final_df

Lot Frontage  Lot Area  ...  Sale Condition_Normal  Sale Condition_Partial
0       141.000000     31770  ...                      1                       0
1        80.000000     11622  ...                      1                       0
2        81.000000     14267  ...                      1                       0
3        93.000000     11160  ...                      1                       0
4        74.000000     13830  ...                      1                       0
...            ...       ...  ...                    ...                     ...
2920     37.000000      7937  ...                      1                       0
2921     75.144444      8885  ...                      1                       0
2922     62.000000     10441  ...                      1                       0
2923     77.000000     10010  ...                      1                       0
2924     74.000000      9627  ...                      1                       0

[2925 rows x 274 columns]

## Réflexions finales

Gardez à l'esprit que nous ne savons pas si avoir 274 colonnes est très utile. Plus de colonnes ne conduit pas nécessairement à de meilleurs résultats. En fait, nous pourrions vouloir supprimer davantage de colonnes (ou plus tard, utiliser un modèle avec régularisation pour choisir les colonnes importantes pour nous). Ce que nous avons fait ici a considérablement augmenté le rapport entre les lignes et les colonnes, ce qui peut en fait conduire à de moins bonnes performances (cependant, vous ne le savez pas tant que vous n'avez pas réellement comparé plusieurs modèles ou approches).

In [ ]:
final_df.corr()['SalePrice'].sort_values()

Exter Qual_TA       -0.591459
Kitchen Qual_TA     -0.527461
Fireplace Qu_None   -0.481740
Bsmt Qual_TA        -0.453022
Garage Finish_Unf   -0.422363
                       ...   
Garage Cars          0.648488
Total Bsmt SF        0.660983
Gr Liv Area          0.727279
Overall Qual         0.802637
SalePrice            1.000000
Name: SalePrice, Length: 274, dtype: float64

    OverallQual: Rates the overall material and finish of the house

           10	Very Excellent
           9	Excellent
           8	Very Good
           7	Good
           6	Above Average
           5	Average
           4	Below Average
           3	Fair
           2	Poor
           1	Very Poor

Il est fort probable qu'un agent immobilier ait évalué cette colonne "Overall Qual", ce qui signifie qu'elle prend très probablement en compte de nombreuses autres features. Cela signifie également que toute maison future pour laquelle nous avons l'intention de prédire un prix aura besoin de cette feature "Overall Qual", ce qui implique que chaque nouvelle maison sur le marché qui sera évaluée avec notre modèle ML nécessitera toujours une personne humaine en amont !

### Sauvegarde du DataFrame final !

In [ ]:
final_df.to_csv('AMES_Final_DF.csv')